tests/lib/timespec_util/src/main.c

```diff
@@ -268,54 +264,152 @@ ZTEST(timeutil_api, test_timespec_equal)
+
+#if defined(CONFIG_TIMEOUT_64BIT) && (CONFIG_SYS_CLOCK_TICKS_PER_SEC == 100)
+	zexpect_equal(K_TS_MAX.tv_nsec, 70000000L);
+#endif
+
+#if (CONFIG_SYS_CLOCK_TICKS_PER_SEC == 32768)
+#if defined(CONFIG_TIMEOUT_64BIT)
+	zexpect_equal(K_TS_MAX.tv_nsec, 999969482L);
+#else
+	zexpect_equal(K_TS_MAX.tv_nsec, 999938964L);
+#endif
+#endif
+}
+
+/* non-saturating */
+#define DECL_TOSPEC_TEST(to, ts, sat, neg, round)                                                  \
+	{                                                                                          \
+		.timeout = (to),                                                                   \
+		.tspec = (ts),                                                                     \
+		.saturation = (sat),                                                               \
+		.negative = (neg),                                                                 \
+		.roundup = (round),                                                                \
+	}
+/* negative timespecs rounded up to K_NO_WAIT */
+#define DECL_TOSPEC_NEGATIVE_TEST(ts)  DECL_TOSPEC_TEST(K_NO_WAIT, (ts), 0, true, false)
+/* zero-valued timeout */
+#define DECL_TOSPEC_ZERO_TEST(to)      DECL_TOSPEC_TEST((to), K_TIMESPEC(0, 0), 0, false, false)
+/* round up toward K_TICK_MIN */
+#define DECL_NSAT_TOSPEC_TEST(ts)      DECL_TOSPEC_TEST(K_TICKS(K_TICK_MIN), (ts), -1, false, false)
+/* round up toward next tick boundary */
+#define DECL_ROUND_TOSPEC_TEST(to, ts) DECL_TOSPEC_TEST((to), (ts), 0, false, true)
+/* round down toward K_TICK_MAX */
+#define DECL_PSAT_TOSPEC_TEST(ts)      DECL_TOSPEC_TEST(K_TICKS(K_TICK_MAX), (ts), 1, false, false)
 
 static const struct tospec {
 	k_timeout_t timeout;
 	struct timespec tspec;
 	int saturation;
+	bool negative;
+	bool roundup;
 } tospecs[] = {
-	{K_NO_WAIT, {INT64_MIN, 0}, -1},
-	{K_NO_WAIT, {-1, 0}, -1},
-	{K_NO_WAIT, {-1, NSEC_PER_SEC - 1}, -1},
-	{K_NO_WAIT, {0, 0}, 0},
-	{K_NSEC(0), {0, 0}, 0},
-	{K_NSEC(2000000000), {2, 0}, 0},
-	{K_USEC(0), {0, 0}, 0},
-	{K_USEC(2000000), {2, 0}, 0},
-	{K_MSEC(100), {0, 100000000}, 0},
-	{K_MSEC(2000), {2, 0}, 0},
-	{K_SECONDS(0), {0, 0}, 0},
-	{K_SECONDS(1), {1, 0}, 0},
-	{K_SECONDS(100), {100, 0}, 0},
-	{K_FOREVER, {INT64_MAX, NSEC_PER_SEC - 1}, 0},
+	/* negative timespecs should round-up to K_NO_WAIT */
+	DECL_TOSPEC_NEGATIVE_TEST(K_TIMESPEC(SYS_TIME_T_MIN, 0)),
+	DECL_TOSPEC_NEGATIVE_TEST(K_TIMESPEC(-1, 0)),
+	DECL_TOSPEC_NEGATIVE_TEST(K_TIMESPEC(-1, NSEC_PER_SEC - 1)),
+
+	/* zero-valued timeouts are equivalent to K_NO_WAIT */
+	DECL_TOSPEC_ZERO_TEST(K_NSEC(0)),
+	DECL_TOSPEC_ZERO_TEST(K_USEC(0)),
+	DECL_TOSPEC_ZERO_TEST(K_MSEC(0)),
+	DECL_TOSPEC_ZERO_TEST(K_SECONDS(0)),
+
+	/* round up to K_TICK_MIN */
+	DECL_NSAT_TOSPEC_TEST(K_TIMESPEC(0, 1)),
+	DECL_NSAT_TOSPEC_TEST(K_TIMESPEC(0, 2)),
+	DECL_NSAT_TOSPEC_TEST(K_TIMESPEC(0, K_TICKS_TO_NSECS(K_TICK_MIN))),
+
+#if CONFIG_SYS_CLOCK_TICKS_PER_SEC < GHZ(1)
+	DECL_NSAT_TOSPEC_TEST(K_TIMESPEC(0, 1)),
```

Comment:
This line is a rep of 348? so I guess instead 348 and 349 should be here?

Reply:
I think that the GHZ condition (and likely the macro definition) can probably be removed. It's highly unlikely that we'll ever have a tick rate that's >= 1 GHz

Reply:
Fixed!